syntax = "proto3";
package school;

message Empty {}

message Student {
  int64 id = 1;
  string name = 2;
}

message Teacher {
  int64 id = 1;
  string name = 2;
}

message Subject {
  int64 id = 1;
  string name = 2;
}

message Class {
  int64 id = 1;
  int64 teacher_id = 2;
  int64 subject_id = 3;
  repeated int64 student_ids = 4;
  string schedule = 5; // exemplo: "Seg 08:00-10:00"
}

message Id {
  int64 id = 1;
}

message EnrollRequest {
  int64 class_id = 1;
  int64 student_id = 2;
}

message StudentList { repeated Student students = 1; }
message TeacherList { repeated Teacher teachers = 1; }
message SubjectList { repeated Subject subjects = 1; }
message ClassList { repeated Class classes = 1; }

service SchoolService {
  // Students
  rpc CreateStudent(Student) returns (Student);
  rpc GetStudent(Id) returns (Student);
  rpc ListStudents(Empty) returns (StudentList);
  rpc DeleteStudent(Id) returns (Empty);

  // Teachers
  rpc CreateTeacher(Teacher) returns (Teacher);
  rpc ListTeachers(Empty) returns (TeacherList);

  // Subjects
  rpc CreateSubject(Subject) returns (Subject);
  rpc ListSubjects(Empty) returns (SubjectList);

  // Classes
  rpc CreateClass(Class) returns (Class);
  rpc ListClasses(Empty) returns (ClassList);
  rpc GetClassesByTeacher(Teacher) returns (ClassList);

  // Enrollment: agora recebe EnrollRequest
  rpc EnrollStudentInClass(EnrollRequest) returns (Class);
}